### PR TITLE
 Merge from upstream through 2021-12-15

### PR DESCRIPTION
**Pull request by amgleitman**
#### Please select one of the following
- [x] I am removing an existing difference between facebook/react-native and microsoft/react-native-macos :thumbsup:
- [ ] I am cherry-picking a change from Facebook's react-native into microsoft/react-native-macos :thumbsup:
- [ ] I am making a fix / change for the macOS implementation of react-native
- [ ] I am making a change required for Microsoft usage of react-native

## Summary

Merge facebook/react-native's main branch into react-native-macos's main branch through 2021-12-15 23:59:59 PST (specifically, 7dfb08a785ae25b60c9344520c679228bdb4f921).

These changes were made by repeatedly running `git merge` and resolving merge conflicts as needed.

## Notes

* 49a1460a379 appears to do a fair amount on iOS, but this shouldn't be relevant to macOS since the latter doesn't have to deal with the issue of an on-screen keyboard obstructing part of the screen.
* baded0f9e3e causes problems with RNTester here because it depends on Fabric, which we don't have for RNTester yet. Mitigated with b85d0db96ec1a6fbc393c3e25b3b51821bdd9206.
* 9827b4ad breaks RNTester when built without Fabric. (I even confirmed this upstream at ea7c9f2a.) The problem is that even without Fabric, we still need `react-native-codegen`. Mitigated with 1175372ffecf6b4e18fa27b75b68f3ec7967a384.

## Validation

* yarn lint
* * ✖ 12 problems (0 errors, 12 warnings)
* yarn flow-check-ios
* yarn flow-check-macos
* yarn flow-check-android
* yarn test
* * Test Suites: 1 skipped, 161 passed, 162 total
* * Tests: 2 skipped, 2319 passed, 2321 total
* * Snapshots: 898 passed, 898 total
* RNTester
* * Compiles on macOS and iOS
* * All unit tests within Xcode pass except for WebSocketTest (this is expected for now) and some snapshot tests on macOS (known issue, see #1184)
* * No new regressions have been found

### CODE REVIEW

File: .github/workflows/create-new-tag-version-label.yml [hunk 1]
```diff
@@ -1,16 +0,0 @@
-name: Create new tag as version label
```
Comment (harrieshin) [inline .github/workflows/create-new-tag-version-label.yml:1]: deleting of this yml file is intentional?

Comment (amgleitman) [reply .github/workflows/create-new-tag-version-label.yml:1]: Yes, it was deleted as part of 63ec8efe58b4a4d768ef065239acbb9c5c67b6fc.

**Comment by dannyvv** [comment]
Not 100% familiar with the code.
Noticed some switches from shared_lib to static_lib and JSI and Hermes changes. But I don't believe that should matter for android.
Deferring to @shivenmian  for better review.

**Comment by dannyvv** [comment]
Note it seems hermes-inspector is nolonger included by default.

---
In reply to: [1150417566](https://github.com/microsoft/react-native-macos/pull/1191#issuecomment-1150417566) [](http://example.com/codeflow?ancestors=1150417566)

**Review by shivenmian — APPROVED** [review]

**Pull request merged**